src/bundles/files.js

```diff
@@ -262,10 +340,51 @@ export default {
     }
   },
 
-  doGetDownloadLink: (files) => async ({ dispatch, store, getIpfs }) => {
-    const ipfs = getIpfs()
-    dispatch({ type: 'FILES_GET_DOWNLOAD_LINK' })
-    return getDownloadLink(files, ipfs)
+  doDownloadFile: (id, hash) => async ({ dispatch, getIpfs }) => {
+    return new Promise((resolve, reject) => {
+      const ipfs = getIpfs()
+      dispatch({ type: 'FILES_DOWNLOAD_STARTED', payload: { id: id } })
+
+      try {
+        const stream = ipfs.files.getReadableStream(hash)
+
+        stream.on('data', (file) => {
+          let chunks = []
+          let bytesLoaded = 0
+
+          file.content
+            .on('data', (chunk) => {
+              bytesLoaded += chunk.byteLength
+              const progress = Math.round((bytesLoaded / file.size) * 100)
+
+              dispatch({ type: 'FILES_DOWNLOAD_PROGRESS', payload: { id: id, progress: progress } })
+              chunks.push(chunk)
+            })
+            .on('end', () => {
```

Comment:
I like that you've done this without adding additional imports, but we could potentially use the Buffer instance on ipfs to concat the chunks in fewer lines (on our side at least) with

```suggestion
            .on('end', () => {
              const buffer = ipfs.types.Buffer.concat(chunks)
              dispatch({ type: 'FILES_DOWNLOAD_FINISHED', payload: { id: id } })
              resolve(buffer)
            })
```

https://github.com/ipfs/js-ipfs-api#domain-data-types
https://github.com/ipfs/js-ipfs#domain-data-types

though, as a type that, I wonder if `window.ipfs` has those...

Reply:
Not yet https://github.com/ipfs-shipyard/ipfs-postmsg-proxy/issues/20

Reply:
I like that. When #20 is done I'll refactor this.